### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by developertariq**
#### Created a class Classroom with the following:
- @label instance variable, initialized in the constructor.
- Setter and getter for @label (remember about attr_accessor).
#### Created the has-many/belongs-to relationship between Classroom and Student. The following implemented:
- Created the has-many side (a classroom has many students).
- Created the belongs-to side (a student belongs to a classroom).
- Confirmed that when adding a student to a classroom it also set the classroom for the student.
- Confirmed that when setting the classroom for a student it also added it to the classroom's students.
#### Created a class Book with the following:
- @title and @author instance variables, initialized in the constructor.
- Setters and getters for instance variables (remember about attr_accessor).
#### Created a class Rental with the following:
- @date instance variable, initialized in the constructor.
- Setter and getter for @date (remember about attr_accessor).
#### Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following implemented:
- Created the has-many side of Book and Rental (a book has many rentals).
- Created the belongs-to side of Rental and Book (a rental belongs to a book).
- Created the has-many side of Person and Rental (a person has many rentals).
- Created the belongs-to side of Rental and Person (a rental belongs to a person).
- Modified the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  def initialize(label)
+    @label = label
+    @students = []
+  end
+
+  def add_student(student)
+    @students << (student)
+    student.classroom = self
+  end
+end
```
Comment (BenMukebo) [inline classroom.rb:13]: - [x]  When setting the classroom for a student it also adds it to the classrooms' students. 👇🏻 


image

- According to the project requirements, 🤔  You should implement a method that serves as the 'belongs-to' side (a student belongs to a classroom) ✍🏻 

- Your code may look like this 👇🏻

```ruby
class Student < Person
  attr_reader :classroom
  .
  def classroom=(classroom)
      ...
  end
end
```

File: rental.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class Rental
+  def initialize(date, book, person)
+    @date = date
+    @book = book
+    @person = person
+    @book.rentals << self
+    @person.rentals << self
+  end
+end
```
Comment (BenMukebo) [inline rental.rb]: - [x] Kindly let's add the setters and getters for the @date instance variable: `attr_accessor :date`.
- [x]  Getters and Setters for @book and @person instance variables should be present as well: ` attr_accessor :date, :book, :person`

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class Book
+  attr_accessor :title, :author, :rentals
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+end
```
Comment (BenMukebo) [inline book.rb]: - [ ] Let's add a method for adding rentals to create the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`. ✍🏻 
- Your implementation might look similar to the following:

```ruby
   .
  def add_rental(.., date)
    ...
  end
```

- Note:  The book itself should be sent as a parameter to create the new rental achieving the 'has-many' association.

File: person.rb [hunk 4]
```diff
@@ -3,7 +3,7 @@
 require_relative 'trimmer'
 
 class Person < Nameable
-  attr_accessor :name, :age
+  attr_accessor :name, :age, :rentals
```
Comment (BenMukebo) [inline person.rb:6]: - [ ] A person must have many rentals. Kindly. Le's add a method for adding rentals  

```ruby
  def add_rental(...)
    ...
  end
```

- Note: The person itself should be sent as a parameter to create the new rental achieving the 'has-many' association

**Review by BenMukebo — CHANGES_REQUESTED** [review]
### Hi @developertariq,

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

### To highlight 💪🏻
- Gitflow was correctly used.  ✔️   
- Good commit messages ✔️ 
- Professional README ✔️ 
 👍  Good job!


## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: rental.rb (hunk 2 above)
Comment (developertariq) [reply rental.rb]: It's done.

Comment (developertariq) [reply rental.rb]: it's done.

File: book.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (OyePriscilla) [inline book.rb]: - [ ] Awesome implementing a method for adding rentals to create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental, I think we can be more specific with the arguments passed in as pointed out by the previous reviewer.

Untitled

```suggestion
 def add_rental(person, date)
    Rental.new(date, self, person)
  end
```
- [optional] Please kindly indicate the `rental` file path that you want to access in line 1 (`require_relative 'rental'`
)

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @developertariq 👏, 

Great work on making the changes requested by a previous reviewer 👏 

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

so-close-so-close-seth-meyers

 ## Highlights
Set-up association between classes and object ✔️ 
Use Ruby Setters and Getters methods ✔️ 


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by bimbolabuari — COMMENTED** [review]
Hello @developertariq 👋,

_This is an additional review as a result of the second opinion request._

- Kudos on working on this milestone 👍 
  - As requested by the previous reviewer, kindly pass the parameters to create (add) new rentals, you will be asked to add book rentals in the upcoming milestone and these parameters `person` and `date` will be used to add new book rentals.
    - You need a person's information and the rental date information in order to save a book rental. 😉 

Check out your `rental.rb`, you can see you have `date`,  `book` and `person` as parameters, in order to instantiate book rentals on `line 8`, you will need the `person`, and `date` parameters in the `book.rb`

https://github.com/developertariq/oop-school-library/blob/02c4cde36bce0272a2b04d40b31d9a0cd78d4e1c/rental.rb#L1-L11

- This clarifies why the previous CRs requested changes.

[Optional] You can also make changes to the `person.rb` file to have the parameter `book` and `date` instead of rentals. 👍 

Keep rocking 🚀!

**Review by alexander16108 — APPROVED** [review]
Hi @developertariq , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

- Gitflow was correctly used. ✔️
- Good commit messages ✔️
- Professional README ✔️

👍 Good job!

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**